evm/src/logic.rs

```diff
@@ -0,0 +1,304 @@
+use crate::stark::Stark;
+use crate::vars::{StarkEvaluationTargets, StarkEvaluationVars};
+
+// Total number of bits per input/output.
+const VAL_BITS: usize = 256;
+// Number of bits stored per field element. Ensure that this fits; it is not checked.
+const PACKED_LIMB_BITS: usize = 16;
+// Number of field elements needed to store each input/output at the specified packing.
+const PACKED_LEN: usize = (VAL_BITS + PACKED_LIMB_BITS - 1) / PACKED_LIMB_BITS;
+// Number of bits in the last limb, which may be smaller than `PACKED_LIMB_BITS`.
+const LAST_LIMB_BITS: usize = VAL_BITS - PACKED_LIMB_BITS * (PACKED_LEN - 1);
+
+pub(crate) mod columns {
+    use std::ops::Range;
+
+    use super::{PACKED_LEN, VAL_BITS};
+
+    pub const IS_AND: usize = 0;
+    pub const IS_OR: usize = IS_AND + 1;
+    pub const IS_XOR: usize = IS_OR + 1;
+    pub const AND_COEFF: usize = IS_XOR + 1;
+    pub const INPUT0_PACKED: Range<usize> = (AND_COEFF + 1)..(AND_COEFF + 1) + PACKED_LEN;
+    pub const INPUT1_PACKED: Range<usize> = INPUT0_PACKED.end..INPUT0_PACKED.end + PACKED_LEN;
+    pub const RESULT: Range<usize> = INPUT1_PACKED.end..INPUT1_PACKED.end + PACKED_LEN;
+    pub const INPUT0_BITS: Range<usize> = RESULT.end..RESULT.end + VAL_BITS;
+    pub const INPUT1_BITS: Range<usize> = INPUT0_BITS.end..INPUT0_BITS.end + VAL_BITS;
+
+    pub const NUM_COLUMNS: usize = INPUT1_BITS.end;
+}
+
+#[derive(Copy, Clone)]
+pub struct LogicStark<F, const D: usize> {
+    pub f: PhantomData<F>,
+}
+
+impl<F: RichField, const D: usize> LogicStark<F, D> {
+    pub fn generate(&self, lv: &mut [F; columns::NUM_COLUMNS]) {
+        let is_and = lv[columns::IS_AND].to_canonical_u64();
+        assert!(is_and <= 1);
+        let is_or = lv[columns::IS_OR].to_canonical_u64();
+        assert!(is_or <= 1);
+        let is_xor = lv[columns::IS_XOR].to_canonical_u64();
+        assert!(is_xor <= 1);
+        assert_eq!(is_and + is_or + is_xor, 1);
+
+        for packed_input_cols in [columns::INPUT0_PACKED, columns::INPUT1_PACKED] {
+            let packed_input = packed_input_cols.map(|i| lv[i].to_canonical_u64());
+            for (i, packed_limb) in packed_input.enumerate() {
+                let bits_in_limb = if i == PACKED_LEN - 1 {
+                    LAST_LIMB_BITS
+                } else {
+                    PACKED_LIMB_BITS
+                };
+                assert_eq!(packed_limb >> bits_in_limb, 0);
+            }
+        }
+
+        for (packed_input_cols, bit_cols) in [
+            (columns::INPUT0_PACKED, columns::INPUT0_BITS),
+            (columns::INPUT1_PACKED, columns::INPUT1_BITS),
+        ] {
+            for (i, limb_col) in packed_input_cols.enumerate() {
+                let limb = lv[limb_col].to_canonical_u64();
+                let limb_bits_cols = bit_cols
+                    .clone()
+                    .skip(i * PACKED_LIMB_BITS)
+                    .take(PACKED_LIMB_BITS);
```

Comment:
nit: might be nice to have a couple `input{0,1}_limb_bits(limb: usize)` methods, since this is a little verbose and we do it a few times

Reply:
Good idea!